File: audit.js
#!/usr/bin/env node
/**
 * Afternoon Plumbing Website Audit
 * Validates: service pages, contact page, CSS, images, navigation, links, structured data, console errors
 */

const fs = require('fs');
const path = require('path');

const BASE_DIR = '/root/.openclaw/workspace/Afternoon-Plumbing-Website';

// Required service pages
const SERVICE_PAGES = [
  'water-softener-installation-westminster-md.html',
  'water-heater-replacement-westminster-md.html',
  'well-pump-repair-westminster-md.html',
  'acid-neutralizer-installation-westminster-md.html',
  'hose-bib-replacement-westminster-md.html'
];

const REQUIRED_PAGES = [...SERVICE_PAGES, 'contact.html', 'index.html'];

// Results collector
const results = {
  pages: {},
  navigation: { consistent: true, issues: [] },
  structuredData: { present: [], missing: [] },
  css: { loaded: true, issues: [] },
  images: { allExist: true, missing: [] },
  internalLinks: { broken: [] },
  consoleErrors: { issues: [] },
  overall: { pass: true, fails: 0 }
};

// Helper: read HTML file
function readHTML(filePath) {
  try {
    return fs.readFileSync(filePath, 'utf8');
  } catch (e) {
    return null;
  }
}

// Helper: extract all href/src URLs from HTML
function extractURLs(html, attr) {
  const regex = new RegExp(`${attr}=["']([^"']+)["']`, 'gi');
  const matches = [];
  let m;
  while ((m = regex.exec(html)) !== null) {
    matches.push(m[1]);
  }
  return matches;
}

// 1. Check all required pages exist and can be read
console.log('\\n🔍 Checking page existence and readability...');
REQUIRED_PAGES.forEach(page => {
  const fullPath = path.join(BASE_DIR, page);
  const exists = fs.existsSync(fullPath);
  const content = exists ? readHTML(fullPath) : null;
  const readable = content && content.includes('<!doctype html');
  results.pages[page] = { exists, readable, missing: !exists || !readable };
  if (!exists || !readable) results.overall.pass = false;
});

// 2. Check navigation consistency
console.log('\\n🔍 Checking navigation consistency...');
const navLinkPatterns = {
  services: /services\.html/,
  faq: /faq\.html/,
  contact: /contact\.html/
};

function extractNavLinks(html) {
  const navMatches = html.match(/<nav[\s\S]*?<\/nav>/i) || [];
  const navText = navMatches.join('');
  const links = {};
  if (navText.includes('Services')) links.services = true;
  if (navText.includes('FAQ')) links.faq = true;
  if (navText.includes('Contact')) links.contact = true;
  return links;
}

REQUIRED_PAGES.forEach(page => {
  const content = results.pages[page].readable ? readHTML(path.join(BASE_DIR, page)) : '';
  if (!content) return;
  const nav = extractNavLinks(content);
  if (!nav.services) results.navigation.issues.push(`${page} missing Services link`);
  if (!nav.faq) results.navigation.issues.push(`${page} missing FAQ link`);
  if (!nav.contact) results.navigation.issues.push(`${page} missing Contact link`);
});

if (results.navigation.issues.length > 0) results.navigation.consistent = false;

// 3. Check structured data on service pages
console.log('\\n🔍 Checking structured data (schema.org)...');
SERVICE_PAGES.forEach(page => {
  const content = readHTML(path.join(BASE_DIR, page));
  if (!content) return;
  const hasLdJson = content.includes('type="application/ld+json"');
  const hasServiceSchema = content.includes('"@type": "Service"') || content.includes('"@type":"Service"');
  if (hasLdJson && hasServiceSchema) {
    results.structuredData.present.push(page);
  } else {
    results.structuredData.missing.push(page);
    results.overall.pass = false;
  }
});

// Also confirm contact and index have schema
['contact.html', 'index.html'].forEach(page => {
  const content = readHTML(path.join(BASE_DIR, page));
  if (content && content.includes('type="application/ld+json"')) {
    results.structuredData.present.push(page);
  }
});

// 4. Check CSS loading
console.log('\\n🔍 Checking CSS loading...');
const cssPath = path.join(BASE_DIR, 'assets', 'index-CDFYR2ld.css');
const cssExists = fs.existsSync(cssPath);
if (!cssExists) {
  results.css.loaded = false;
  results.css.issues.push('Main CSS file not found: assets/index-CDFYR2ld.css');
  results.overall.pass = false;
}

// Verify CSS is referenced in pages
REQUIRED_PAGES.forEach(page => {
  const content = readHTML(path.join(BASE_DIR, page));
  if (content && content.includes('index-CDFYR2ld.css')) {
    // OK
  } else if (content) {
    results.css.issues.push(`${page} does not reference main CSS`);
    results.overall.pass = false;
  }
});

// 5. Check all images referenced exist in /images/
console.log('\\n🔍 Checking image references...');
const imageFiles = fs.readdirSync(path.join(BASE_DIR, 'images'));
const imageSet = new Set(imageFiles.map(f => f.toLowerCase()));

function extractImageURLs(html) {
  const urls = extractURLs(html, 'src');
  const imgUrls = urls.filter(u => u.includes('/images/') || u.startsWith('/images/') || u.includes('images/'));
  return imgUrls;
}

let allImagesFound = true;
REQUIRED_PAGES.forEach(page => {
  const content = readHTML(path.join(BASE_DIR, page));
  if (!content) return;
  const imgUrls = extractImageURLs(content);
  imgUrls.forEach(url => {
    const filename = url.split('/').pop().split('?')[0].toLowerCase();
    if (!imageSet.has(filename)) {
      results.images.missing.push({ page, image: filename });
      allImagesFound = false;
    }
  });
});
results.images.allExist = allImagesFound;
if (!allImagesFound) results.overall.pass = false;

// 6. Check for broken internal links (same-site absolute or relative)
console.log('\\n🔍 Checking internal links...');
const internalLinkRegex = /href=["'](\/([^"']+\.html)|(#[^"']+)|([^"']+\.html))["']/gi;
let m;
REQUIRED_PAGES.forEach(page => {
  const content = readHTML(path.join(BASE_DIR, page));
  if (!content) return;
  while ((m = internalLinkRegex.exec(content)) !== null) {
    let href = m[1];
    if (!href) continue;
    // Skip anchors
    if (href.startsWith('#')) continue;
    // Skip tel: and mailto:
    if (href.startsWith('tel:') || href.startsWith('mailto:')) continue;
    // Resolve relative paths
    let targetPath = href;
    if (!href.startsWith('/')) {
      const pageDir = path.dirname(page);
      targetPath = path.join(pageDir, href);
    }
    // Normalize
    targetPath = targetPath.replace(/\/\/+/g, '/').replace(/^\//, '');
    const fullTarget = path.join(BASE_DIR, targetPath);
    if (!fs.existsSync(fullTarget)) {
      results.internalLinks.broken.push({ from: page, to: href });
      results.overall.pass = false;
    }
  }
});

// 7. Check contact page for email button (mailto link)
console.log('\\n🔍 Checking email button on contact page...');
const contactContent = readHTML(path.join(BASE_DIR, 'contact.html'));
const hasEmailButton = contactContent && contactContent.includes('mailto:afternoonplumbingcc@gmail.com');
results.contactEmailButton = hasEmailButton;
if (!hasEmailButton) {
  results.overall.pass = false;
}

// 8. Check for JavaScript console errors patterns (like common JS errors)
console.log('\\n🔍 Checking for potential console errors in scripts...');
const jsErrorPatterns = [
  /console\.error/gi,
  /throw new Error/gi,
  /uncaught/i,
  /ReferenceError/gi,
  /TypeError/gi
];
REQUIRED_PAGES.forEach(page => {
  const content = readHTML(path.join(BASE_DIR, page));
  if (!content) return;
  jsErrorPatterns.forEach(pattern => {
    if (pattern.test(content)) {
      results.consoleErrors.issues.push(`${page} contains potential console error code`);
    }
  });
});

// Summary
console.log('\\n' + '='.repeat(60));
console.log('AUDIT REPORT: Afternoon Plumbing Website');
console.log('='.repeat(60));

function passFail(status) {
  return status ? '✅ PASS' : '❌ FAIL';
}

console.log(`\\n📄 Page Existence & Validity:`);
REQUIRED_PAGES.forEach(page => {
  const p = results.pages[page];
  console.log(`  ${passFail(p.exists && p.readable)} ${page}`);
});

console.log(`\\n🧭 Navigation Consistency:`);
console.log(`  ${passFail(results.navigation.consistent)} All pages include Services, FAQ, Contact links`);
if (results.navigation.issues.length) {
  results.navigation.issues.forEach(issue => console.log(`    ⚠️ ${issue}`));
}

console.log(`\\n🔎 Structured Data (schema.org):`);
SERVICE_PAGES.forEach(page => {
  const has = results.structuredData.present.includes(page);
  console.log(`  ${passFail(has)} ${page} (Service)`);
});
['contact.html', 'index.html'].forEach(page => {
  const has = results.structuredData.present.includes(page);
  console.log(`  ${passFail(has)} ${page} (Other)`);
});

console.log(`\\n🎨 CSS Loading:`);
console.log(`  ${passFail(results.css.loaded)} assets/index-CDFYR2ld.css exists`);
if (results.css.issues.length) {
  results.css.issues.forEach(issue => console.log(`    ⚠️ ${issue}`));
}

console.log(`\\n🖼️ Images:`);
console.log(`  ${passFail(results.images.allExist)} All referenced images exist in /images/`);
if (!results.images.allExist) {
  results.images.missing.forEach(m => {
    console.log(`    ⚠️ ${m.page}: missing ${m.image}`);
  });
}

console.log(`\\n🔗 Internal Links:`);
console.log(`  ${passFail(results.internalLinks.broken.length === 0)} No broken internal links`);
if (results.internalLinks.broken.length) {
  results.internalLinks.broken.forEach(l => {
    console.log(`    ⚠️ ${l.from} -> ${l.to} (broken)`);
  });
}

console.log(`\\n📧 Contact Email Button:`);
console.log(`  ${passFail(results.contactEmailButton)} contact.html contains email button (mailto)`);

console.log(`\\n⚠️ Console Errors Check:`);
console.log(`  ${passFail(results.consoleErrors.issues.length === 0)} No obvious JS error patterns found`);
if (results.consoleErrors.issues.length) {
  results.consoleErrors.issues.forEach(i => console.log(`    ⚠️ ${i}`));
}

console.log('\\n' + '='.repeat(60));
console.log(`\\n🏆 OVERALL RESULT: ${results.overall.pass ? '✅ PASS' : '❌ FAIL'}`);
console.log('='.repeat(60));

// Exit with appropriate code
process.exit(results.overall.pass ? 0 : 1);

File: audit.py
#!/usr/bin/env python3
"""Afternoon Plumbing Website Audit"""

import os
import re
from pathlib import Path

BASE_DIR = Path('/root/.openclaw/workspace/Afternoon-Plumbing-Website')

# Required pages
SERVICE_PAGES = [
    'water-softener-installation-westminster-md.html',
    'water-heater-replacement-westminster-md.html',
    'well-pump-repair-westminster-md.html',
    'acid-neutralizer-installation-westminster-md.html',
    'hose-bib-replacement-westminster-md.html'
]
REQUIRED_PAGES = SERVICE_PAGES + ['contact.html', 'index.html']

results = {
    'pages': {},
    'navigation_issues': [],
    'structured_missing': [],
    'css_issues': [],
    'images_missing': [],
    'broken_links': [],
    'email_button': False,
    'console_issues': []
}

def read_html(page):
    path = BASE_DIR / page
    if not path.exists():
        return None
    try:
        return path.read_text(encoding='utf-8', errors='ignore')
    except:
        return None

print('='*60)
print('WEBSITE AUDIT: Afternoon Plumbing')
print('='*60)

# 1. Page existence
print('\n📄 Page Existence & Readability:')
for page in REQUIRED_PAGES:
    content = read_html(page)
    if content and '<!doctype html>' in content.lower():
        results['pages'][page] = True
        print(f'✅ PASS: {page}')
    else:
        results['pages'][page] = False
        print(f'❌ FAIL: {page} missing or invalid')

# 2. Navigation consistency
print('\n🧭 Navigation Consistency:')
for page in REQUIRED_PAGES:
    content = read_html(page)
    if not content:
        continue
    issues = []
    if 'services.html' not in content:
        issues.append('Services')
    if 'faq.html' not in content:
        issues.append('FAQ')
    if 'contact.html' not in content:
        issues.append('Contact')
    if issues:
        results['navigation_issues'].append(f'{page} missing: {", ".join(issues)}')
        print(f'  ⚠️  {page} missing links: {", ".join(issues)}')
if not results['navigation_issues']:
    print('✅ PASS: All pages include Services, FAQ, Contact links')
else:
    print('❌ FAIL: Some pages missing navigation links')

# 3. Structured data
print('\n🔎 Structured Data (schema.org):')
for page in SERVICE_PAGES:
    content = read_html(page)
    if content and 'type="application/ld+json"' in content and '"@type": "Service"' in content:
        print(f'✅ PASS: {page} has Service schema')
    else:
        results['structured_missing'].append(page)
        print(f'❌ FAIL: {page} missing Service schema')
for page in ['contact.html', 'index.html']:
    content = read_html(page)
    if content and 'type="application/ld+json"' in content:
        print(f'✅ PASS: {page} has structured data')
    else:
        results['structured_missing'].append(page)
        print(f'❌ FAIL: {page} missing structured data')

# 4. CSS
print('\n🎨 CSS Loading:')
# Index uses external CSS
index_content = read_html('index.html')
if index_content and 'index-CDFYR2ld.css' in index_content:
    print('✅ PASS: index.html references main CSS')
else:
    print('❌ FAIL: index.html missing main CSS reference')
# Contact and service pages use embedded CSS
for page in ['contact.html'] + SERVICE_PAGES:
    content = read_html(page)
    if content and '<style>' in content:
        print(f'✅ PASS: {page} has embedded CSS')
    else:
        print(f'❌ FAIL: {page} missing CSS')

# 5. Images
print('\n🖼️ Images:')
# Build set of existing images (lowercase)
image_dir = BASE_DIR / 'images'
existing_images = set(f.name.lower() for f in image_dir.iterdir() if f.is_file())
missing_images = set()
for page in REQUIRED_PAGES:
    content = read_html(page)
    if not content:
        continue
    # Find all /images/FILENAME references
    for match in re.finditer(r'src=["\']?/images/([^"\'>\s]+)', content):
        img = match.group(1).lower()
        if img not in existing_images:
            missing_images.add(img)
            print(f'  ⚠️  {page} references missing: /images/{img}')
if not missing_images:
    print('✅ PASS: All referenced images exist')
else:
    print(f'❌ FAIL: {len(missing_images)} missing image(s)')

# 6. Internal links
print('\n🔗 Internal Links:')
broken_links = []
for page in REQUIRED_PAGES + ['services.html', 'faq.html']:
    content = read_html(page)
    if not content:
        continue
    # Find href="something.html" (not external, tel, mailto, #anchor)
    for match in re.finditer(r'href=["\']([^"\'#]+\.html)["\']', content):
        href = match.group(1)
        # Skip external
        if href.startswith('http://') or href.startswith('https://'):
            continue
        # Skip tel/mailto
        if href.startswith('tel:') or href.startswith('mailto:'):
            continue
        # Resolve path
        if href.startswith('/'):
            target = BASE_DIR / href[1:]
        else:
            target = (BASE_DIR / page).parent / href
        target = target.resolve()
        if not target.exists():
            broken_links.append(f'{page} -> {href}')
            print(f'  ⚠️  {page} -> {href} (broken)')
if not broken_links:
    print('✅ PASS: No broken internal links')
else:
    print(f'❌ FAIL: {len(broken_links)} broken link(s)')

# 7. Contact email button
print('\n📧 Contact Email Button:')
contact_content = read_html('contact.html')
if contact_content and 'mailto:afternoonplumbingcc@gmail.com' in contact_content:
    print('✅ PASS: contact.html has email button (mailto)')
else:
    print('❌ FAIL: contact.html missing email button')

# 8. Console error patterns
print('\n⚠️  Console Error Patterns:')
console_issues = 0
for page in REQUIRED_PAGES:
    content = read_html(page)
    if content and re.search(r'console\.error|throw new Error|uncaught|ReferenceError|TypeError', content):
        print(f'  ⚠️  {page} contains potential error code')
        console_issues += 1
if console_issues == 0:
    print('✅ PASS: No obvious JS error patterns found')
else:
    print(f'ℹ️  {console_issues} page(s) with potential errors')

# Summary
print('\n' + '='*60)
print('SUMMARY')
print('='*60)
# Count checks
checks = {
    'pages': all(results['pages'].values()),
    'navigation': len(results['navigation_issues']) == 0,
    'structured': len(results['structured_missing']) == 0,
    'css': len(results['css_issues']) == 0,
    'images': len(missing_images) == 0,
    'links': len(broken_links) == 0,
    'email': contact_content and 'mailto:afternoonplumbingcc@gmail.com' in contact_content,
    'console': console_issues == 0
}
passed = sum(1 for v in checks.values() if v)
failed = sum(1 for v in checks.values() if not v)
print(f'✅ Passed: {passed}/8')
print(f'❌ Failed: {failed}/8')
print()
if failed == 0:
    print('🏆 OVERALL RESULT: ✅ PASS')
    exit(0)
else:
    print('🏆 OVERALL RESULT: ❌ FAIL')
    exit(1)
